src/snowflake/cli/plugins/workspace/manager.py

```diff
@@ -0,0 +1,42 @@
+from pathlib import Path
+from typing import Dict
+
+from snowflake.cli.api.exceptions import InvalidProjectDefinitionVersionError
+from snowflake.cli.api.project.schemas.entities.entities import (
+    Entity,
+    v2_entity_model_to_entity_map,
+)
+from snowflake.cli.api.project.schemas.project_definition import DefinitionV20
+
+
+class WorkspaceManager:
+    """
+    Instantiates entity instances from entity models, providing higher-order functionality on entity compositions.
+    """
+
+    def __init__(self, project_definition: DefinitionV20, project_root: Path):
+        if (
+            project_definition.definition_version != "2"
+            and not project_definition.definition_version.startswith("2.")
+        ):
+            raise InvalidProjectDefinitionVersionError(
+                "2.x", project_definition.definition_version
+            )
+        self._entities_cache: Dict[str, Entity] = {}
+        self._project_definition = project_definition
+
+    def get_entity(self, entity_id: str):
+        if entity_id not in self._entities_cache:
+            if entity_id not in self._project_definition.entities:
+                raise ValueError(f"No such entity ID: {entity_id}")
+            entity_model_cls = self._project_definition.entities[entity_id].__class__
+            entity_cls = v2_entity_model_to_entity_map[entity_model_cls]
+            self._entities_cache[entity_id] = entity_cls()
+        return self._entities_cache[entity_id]
+
+    def bundle(self, entity_id: str):
+        entity = self.get_entity(entity_id)
+        if callable(getattr(entity, "bundle", None)):
```

Comment:
@sfc-gh-turbaszek would love to get your thoughts on the way you'd like to dispatch things here. You had suggested a decorator-based approach, e.g.

```python
@action(type=Actions.BUNDLE)
def bundle(self):
    [...]
```

After thinking about it, I realize I don't have strong enough feelings either way, except for if we go for this approach I'd like to encapsulate it, e.g.

```suggestion
        if entity.supports(Action.BUNDLE):
```

Reply:
My 2 cents:
1. Imho the method have to be decorated as action. We should not assume every instance attribute can be cli action. Especially than not every attribute has to be callable
2. +1 for `entity.supports(Action.BUNDLE)`
4. Why are we implementing `bundle` method here? It shouldn't it be `perform_action(entity_name, action_name)`?

Reply:
Changed to the more generic method `perform_action(entity_id, action)` and added `entity.supports(action)`.

For now an entity is considered to support an action if it implements a method with the action name. Will refactor this convention once we introduce the action decorator.

Reply:
Let's add this comment in the code, so others also know this.